pallets/parachain-staking/src/lib.rs

```diff
@@ -0,0 +1,1870 @@
+                if top_n > T::MinSelectedCandidates::get() as usize {
+                    top_n -= 1;
+                }
+            }
+            // END MANTA WORKAROUND
+            // choose the top TotalSelected qualified candidates, ordered by stake
+            let mut collators = candidates
+                .into_iter()
+                .rev()
+                .take(top_n)
+                .filter(|x| x.amount >= T::MinCollatorStk::get())
+                .map(|x| x.owner)
+                .collect::<Vec<T::AccountId>>();
+            collators.sort();
+            collators
+        }
+        /// Best as in most cumulatively supported in terms of stake
+        /// Returns [collator_count, delegation_count, total staked]
+        fn select_top_candidates(now: RoundIndex) -> (u32, u32, BalanceOf<T>) {
+            let (mut collator_count, mut delegation_count, mut total) =
+                (0u32, 0u32, BalanceOf::<T>::zero());
+            // choose the top TotalSelected qualified candidates, ordered by stake
+            let collators = Self::compute_top_candidates();
+            if collators.is_empty() {
+                // SELECTION FAILED TO SELECT >=1 COLLATOR => select collators from previous round
+                let last_round = now.saturating_sub(1u32);
+                let mut total_per_candidate: BTreeMap<T::AccountId, BalanceOf<T>> = BTreeMap::new();
+                // set this round AtStake to last round AtStake
+                for (account, snapshot) in <AtStake<T>>::iter_prefix(last_round) {
+                    collator_count = collator_count.saturating_add(1u32);
+                    delegation_count =
+                        delegation_count.saturating_add(snapshot.delegations.len() as u32);
+                    total = total.saturating_add(snapshot.total);
+                    total_per_candidate.insert(account.clone(), snapshot.total);
+                    <AtStake<T>>::insert(now, account, snapshot);
+                }
+                // `SelectedCandidates` remains unchanged from last round
+                // emit CollatorChosen event for tools that use this event
+                for candidate in <SelectedCandidates<T>>::get() {
+                    let snapshot_total = total_per_candidate
+                        .get(&candidate)
+                        .expect("all selected candidates have snapshots");
+                    Self::deposit_event(Event::CollatorChosen {
+                        round: now,
+                        collator_account: candidate,
+                        total_exposed_amount: *snapshot_total,
+                    })
+                }
+                return (collator_count, delegation_count, total);
+            }
+
+            // snapshot exposure for round for weighting reward distribution
+            for account in collators.iter() {
+                let state = <CandidateInfo<T>>::get(account)
+                    .expect("all members of CandidateQ must be candidates");
```

Comment:
expect on storage

Reply:
compute_top_candidates() returns candidates, so this work